File: libgdx_galaxy_invaders/core/src/com/nopalsoft/invaders/frame/Missile.java
package com.nopalsoft.invaders.frame;

import com.badlogic.gdx.math.Circle;

public class Missile extends DynamicGameObject {

    public static final float WIDTH = 0.4f;
    public static final float HEIGHT = 1.4f;

    public static final float RADIO_EXPLOSION = 7.5f;
    public static final float TIEMPO_EXPLODE = 0.05f * 19;
    public final static int STATE_DISPARADO = 0;
    public final static int STATE_EXPLOTANDO = 1;
    public final float VELOCIDAD = 30;
    public float stateTime;
    public int state;

    /**
     * X y Y son la posicion de la punta de nave
     *
     * @param x La misma que Bob.x
     * @param y La misma que Bob.y
     */
    public Missile(float x, float y) {
        super(x, y, WIDTH, HEIGHT);
        // Tambien inicializo el radio porque la explosion va estar guapetona
        boundsCircle = new Circle(position.x, position.y, RADIO_EXPLOSION);
        state = STATE_DISPARADO;
        stateTime = 0;
        velocity.set(0, VELOCIDAD);
    }

    public void update(float deltaTime) {
        position.add(velocity.x * deltaTime, velocity.y * deltaTime);
        boundsRectangle.x = position.x - WIDTH / 2;
        boundsRectangle.y = position.y - HEIGHT / 2;
        boundsCircle.x = position.x;
        boundsCircle.y = position.y;
        stateTime += deltaTime;
    }

    public void hitTarget() {
        velocity.set(0, 0);
        stateTime = 0;
        state = STATE_EXPLOTANDO;
    }

}

File: libgdx_galaxy_invaders/core/src/com/nopalsoft/invaders/frame/Missile.kt
package com.nopalsoft.invaders.frame

import com.badlogic.gdx.math.Circle


/**
 * X and Y are the position of the tip of the ship.
 *
 * @param x The same as Bob.x
 * @param y The same as Bob.y
 */
class Missile(x: Float, y: Float) : DynamicGameObject(x, y, WIDTH, HEIGHT) {

    private val speed: Float = 30f

    var stateTime: Float
    var state: Int

    init {
        // I also initialize the radius because the explosion will be nicer.
        boundsCircle = Circle(position.x, position.y, RADIO_EXPLOSION)
        state = STATE_TRIGGERED
        stateTime = 0f
        velocity[0f] = speed
    }

    fun update(deltaTime: Float) {
        position.add(velocity.x * deltaTime, velocity.y * deltaTime)
        boundsRectangle!!.x = position.x - WIDTH / 2
        boundsRectangle.y = position.y - HEIGHT / 2
        boundsCircle!!.x = position.x
        boundsCircle!!.y = position.y
        stateTime += deltaTime
    }

    fun hitTarget() {
        velocity[0f] = 0f
        stateTime = 0f
        state = STATE_EXPLODING
    }

    companion object {
        const val WIDTH: Float = 0.4f
        const val HEIGHT: Float = 1.4f

        const val RADIO_EXPLOSION = 7.5f
        const val EXPLODE_TIME = 0.05f * 19
        const val STATE_TRIGGERED = 0
        const val STATE_EXPLODING = 1
    }
}
